### PR TITLE
feat(llma): (3/4) Add HDBSCAN clustering workflow with LangGraph labeling agent

### PR DESCRIPTION
**Pull request by andrewm4894**
This PR adds some improvements and flexibility to the temporal trace clustering workflow. Also adds a labeling agent for the cluster label step as its much better than the one shot llm approach we were initially using. Add's more params and flexibility too about the inputs and ML recipe part of the clustering job.

## Problem

LLM Analytics users need a way to discover patterns in their traces. Without clustering, users have to manually sift through individual traces to find common patterns, errors, or usage behaviors.

Part 3 of 4 PRs split from #42608 (see [split plan](https://github.com/PostHog/posthog/pull/42608#issuecomment-3657905829)).

Depends on #43487 (trace summarization enhancements).

## Changes

- Add umap-learn dependency for dimensionality reduction (3072D -> 100D for clustering, 2D for visualization)
- Replace k-means with HDBSCAN density-based clustering for automatic cluster count detection
- Add LangGraph labeling agent with Claude Sonnet 4 for generating distinctive cluster labels
- Agent has 6 tools: get_clusters_overview, get_cluster_trace_titles, get_trace_details, get_current_labels, set_cluster_label, finalize_labels
- Support optional trace property filtering for scoped clustering
- Add L2 embedding normalization option
- Reorganize tests into tests/ directory
- Add clustering API endpoint

## How did you test this code?

- Added test_clustering.py, test_data.py, test_event_emission.py
- Ran clustering workflow with 1000 traces, verified clusters generated with distinctive labels
- Tested HDBSCAN and K-means clustering methods
- Tested UMAP and PCA dimensionality reduction options

## Changelog: No

Hidden behind feature flag.

### CODE REVIEW

**Comment by andrewm4894** [comment]
labeling agent here is too low level - going to use some of the higher level langgraph/chain abstractions

File: posthog/temporal/llm_analytics/trace_clustering/constants.py [hunk 1]
```diff
@@ -70,7 +74,18 @@
     112495,  # Dogfooding project
 ]
 
-# Cluster labeling configuration
-DEFAULT_TRACES_PER_CLUSTER_FOR_LABELING = 7  # Number of representative traces to use for LLM labeling
-LABELING_LLM_MODEL = "gpt-5.1"
-LABELING_LLM_TIMEOUT = 240.0
+# Cluster labeling agent configuration
+LABELING_AGENT_MODEL = "claude-sonnet-4-5"  # Claude Sonnet 4.5 for reasoning
+LABELING_AGENT_MAX_ITERATIONS = 50  # Max agent iterations before forced finalization
```
Comment (Radu-Raicea) [inline posthog/temporal/llm_analytics/trace_clustering/constants.py:79]: Do we really need this many iterations? It seems pretty excessive to get to 50 iterations.

File: posthog/temporal/llm_analytics/trace_clustering/data.py [hunk 2]
```diff
@@ -7,109 +7,278 @@
...
+
+    # Build event types tuple for IN clause
+    event_types_tuple = ast.Tuple(exprs=[ast.Constant(value=e) for e in AI_EVENT_TYPES])
+
     query = parse_select(
         """
-        SELECT document_id, embedding
-        FROM raw_document_embeddings
-        WHERE timestamp >= {start_dt}
+        SELECT DISTINCT properties.$ai_trace_id as trace_id
```
Comment (Radu-Raicea) [inline posthog/temporal/llm_analytics/trace_clustering/data.py:73]: Without a team_id, this is a full table scan on the given partitions.

Comment (Radu-Raicea) [reply posthog/temporal/llm_analytics/trace_clustering/data.py:73]: Oh, this is HogQL right? It should include the team_id.

File: products/llm_analytics/backend/api/clustering.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    # Trace filters
+    trace_filters = serializers.ListField(
+        child=serializers.DictField(),
+        required=False,
+        default=list,
+        help_text="Property filters to scope which traces are included in clustering (PostHog standard format)",
+    )
+
+
+class LLMAnalyticsClusteringRunViewSet(TeamAndOrgViewSetMixin, viewsets.ViewSet):
```
Comment (Radu-Raicea) [inline products/llm_analytics/backend/api/clustering.py:144]: Is the endpoint itself behind a feature flag?

File: posthog/temporal/llm_analytics/trace_clustering/clustering.py [hunk 4]
```diff
@@ -82,42 +87,245 @@ def calculate_trace_distances(
...
+        random_state=random_state,
+    )
+
+    reduced = reducer.fit_transform(embeddings)
+    return reduced, reducer
+
+
+def perform_hdbscan_clustering(
+    embeddings: np.ndarray,
+    min_cluster_size_fraction: float = 0.05,
```
Comment (Radu-Raicea) [inline posthog/temporal/llm_analytics/trace_clustering/clustering.py:251]: Should we validate the value for `min_cluster_size_fraction`?

**Review by Radu-Raicea — COMMENTED** [review]
I left a few comments to get to.

Overall, I'm not a fan of this PR because of the following reasons:
- It's huge
- It has many different types of changes that are not necessarily related or required to be done at once
- The changes are **highly** complex
- You provide too many customization options, increasing the complexity

The result of all of those combined is a shallow PR review (which is unfortunately what I could muster), a very slow review, and the knowledge is somewhat siloed to you. If we have bugs with this, it's going to be extremely hard for the rest of the team to figure out what they are, and how to fix them.

I think that, at the very least, the labelling agent could have been a separate PR. Then, you could have separated the new clustering algorithm in a new PR (and perhaps even file, where Kmeans is one file, and HDBSCAN another). The new endpoint to trigger a cluster could have been its own PR too.

File: posthog/temporal/llm_analytics/trace_clustering/constants.py (hunk 1 above)
Comment (andrewm4894) [reply posthog/temporal/llm_analytics/trace_clustering/constants.py:79]: yeah so with the dbscan we can end up with 10-15 clusters and so it does sorta need to sample each cluster a few times. that said its more like an upper bound. 

this is actually a good example of a "metric" i would like to just get from llma in some way so we can just measure it.

im going to think about if we have a way to do this easy out of the box via some custom sql even perhaps and see if can use it as a sort of POC for "agent metrics" that tools like llma should just give you out of the box.

really ultimately its sorta an empirical one would be good to measure the distribution of n_iterations over all teams for example.

File: posthog/temporal/llm_analytics/trace_clustering/data.py (hunk 2 above)
Comment (andrewm4894) [reply posthog/temporal/llm_analytics/trace_clustering/data.py:73]: yep hogql so team id injected in `execute_hogql_query()` automatically

File: products/llm_analytics/backend/api/clustering.py (hunk 3 above)
Comment (andrewm4894) [reply products/llm_analytics/backend/api/clustering.py:144]: good idea - added one [here](https://github.com/PostHog/posthog/pull/43488/commits/f2fd0da42f6c15a3028439966f35b75353abd64d) similar to the prompt one at user level

File: posthog/temporal/llm_analytics/trace_clustering/clustering.py (hunk 4 above)
Comment (andrewm4894) [reply posthog/temporal/llm_analytics/trace_clustering/clustering.py:251]: added into the function here to be defensive - the api already does some validation but no harm checking in the function too - https://github.com/PostHog/posthog/pull/43488/commits/5dc301ec9d758f314914a738b70fbb776d6427f7

File: posthog/temporal/llm_analytics/trace_clustering/constants.py (hunk 1 above)
Comment (Radu-Raicea) [reply posthog/temporal/llm_analytics/trace_clustering/constants.py:79]: I think it's important for us to track costs for these. We have some way of doing that for evals.

File: posthog/temporal/llm_analytics/trace_clustering/clustering.py (hunk 4 above)
Comment (Radu-Raicea) [reply posthog/temporal/llm_analytics/trace_clustering/clustering.py:251]: Code changes over time, the validation in the API could change and this could fail, so it's good to be defensive.

**Review by Radu-Raicea — APPROVED** [review]
